flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java

```diff
@@ -0,0 +1,256 @@
+
+					if (jobStatus != DispatcherJobStatus.CANCELLING) {
+						jobStatus = DispatcherJobStatus.JOB_MANAGER_RUNNER_INITIALIZED;
+					}
+				} else { // failure during initialization
+					if (executionType == Dispatcher.ExecutionType.RECOVERY) {
+						jobResultFuture.completeExceptionally(throwable);
+					} else {
+						jobResultFuture.complete(ArchivedExecutionGraph.createFromInitializingJob(
+							jobId,
+							jobName,
+							JobStatus.FAILED,
+							throwable,
+							initializationTimestamp));
+					}
+				}
+			}
+			return null;
+		}));
+	}
+
+	public CompletableFuture<ArchivedExecutionGraph> getResultFuture() {
+		return jobResultFuture;
+	}
+
+	public CompletableFuture<JobDetails> requestJobDetails(Time timeout) {
+		synchronized (lock) {
+			if (isInitialized()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.requestJobDetails(
+					timeout));
+			} else {
+				return getInitializingJobDetails(timeout);
+			}
+		}
+	}
+
+	public CompletableFuture<Acknowledge> cancel(Time timeout) {
+		synchronized (lock) {
+			if (isInitialized()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.cancel(timeout));
+			} else {
+				log.info("Cancellation during initialization requested for job {}. Job will be cancelled once JobManager has been initialized.", jobId);
+
+				// cancel job
+				jobManagerRunnerFuture
+					.thenCompose(jobManagerRunner -> getJobMasterGateway())
+					.thenCompose(gw -> gw.cancel(RpcUtils.INF_TIMEOUT))
+					.whenComplete((ignored, cancelThrowable) -> {
+						if (cancelThrowable != null) {
+							log.warn("Cancellation of job {} failed", jobId, cancelThrowable);
+						}
+					});
+				jobStatus = DispatcherJobStatus.CANCELLING;
+				jobResultFuture.whenComplete(((archivedExecutionGraph, throwable) -> {
+					if (archivedExecutionGraph != null) {
+						jobStatus = DispatcherJobStatus.JOB_MANAGER_RUNNER_INITIALIZED;
+					}
+				}));
+				return jobResultFuture.thenApply(ignored -> Acknowledge.get());
+			}
+		}
+	}
+
+	public CompletableFuture<JobStatus> requestJobStatus(Time timeout) {
+		return requestJob(timeout).thenApply(ArchivedExecutionGraph::getState);
+	}
+
+	public CompletableFuture<ArchivedExecutionGraph> requestJob(Time timeout) {
+		synchronized (lock) {
+			if (isInitialized()) {
+				if (jobResultFuture.isDone()) { // job is not running anymore
+					return jobResultFuture;
+				}
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.requestJob(
+					timeout));
+			} else {
+				Preconditions.checkState(jobStatus == DispatcherJobStatus.INITIALIZING || jobStatus == DispatcherJobStatus.CANCELLING);
+				return CompletableFuture.supplyAsync(() -> {
+					synchronized (lock) {
+						return ArchivedExecutionGraph.createFromInitializingJob(
+							jobId,
+							jobName,
+							jobStatus.asJobStatus(),
+							null,
+							initializationTimestamp);
+					}
+				});
```

Comment:
Why are you using `supplyAsync` here? `createFromInitializingJob` should be non-blocking.